File: backend/app/schemas/supply_chain.py
from pydantic import BaseModel


class SupplyChainNode(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    relation_type: str        # upstream / downstream / competitor
    company_name: str
    company_code: str | None = None
    product_desc: str | None = None
    percentage: float | None = None
    importance: str           # high / medium / low
    is_listed: bool
    data_source: str
    report_year: int | None = None
    recent_event_count: int = 0
    urgent_event_count: int = 0
    latest_event_title: str | None = None
    latest_news_id: int | None = None


class SupplyChainEventSummary(BaseModel):
    """供应链图谱中的近期重要资讯摘要。"""

    news_id: int
    stock_code: str
    stock_name: str
    company_name: str
    company_code: str | None = None
    title: str
    published_at: str | None = None
    urgency: str | None = None
    importance_score: float | None = None
    relevance: float
    impact_direction: str | None = None
    impact_summary: str | None = None


class SupplyChainRead(BaseModel):
    code: str
    name: str
    upstream: list[SupplyChainNode] = []
    downstream: list[SupplyChainNode] = []
    competitors: list[SupplyChainNode] = []
    updated_at: str | None = None
    recent_events: list[SupplyChainEventSummary] = []


# ── 全局供应链图(聚簇分组) ─────────────────────────────────────
class SupplyChainStockMeta(BaseModel):
    """单只自选股的元数据(用于聚簇分组)"""
    code: str
    name: str
    industry: str | None = None
    market: str = "A"  # A / H
    recent_event_count: int = 0
    urgent_event_count: int = 0
    latest_event_title: str | None = None
    latest_news_id: int | None = None


class SupplyChainEdge(BaseModel):
    """边:from_code → to_code,标注关系类型与产品描述"""
    from_code: str        # 上游公司 code(可能是非自选股的占位 code)
    to_code: str          # 下游公司 code
    from_name: str
    to_name: str
    product_desc: str | None = None
    importance: str       # high / medium / low
    relation_type: str    # upstream / downstream
    # 是否为自选股之间的连接(双向都是自选股则 True,用于高亮)
    both_listed: bool = False
    recent_event_count: int = 0
    urgent_event_count: int = 0
    latest_event_title: str | None = None
    latest_news_id: int | None = None


class GlobalSupplyChainResponse(BaseModel):
    """全局供应链图 — 自选股 + 上下游伙伴的聚簇网络"""
    watchlist_stocks: list[SupplyChainStockMeta]   # 所有自选股
    external_companies: list[SupplyChainStockMeta] # 非自选股的上下游公司
    edges: list[SupplyChainEdge]
    recent_events: list[SupplyChainEventSummary] = []
    industry_groups: dict[str, list[str]]          # industry -> list of codes(用于聚簇)
    stats: dict[str, int]                          # 节点/边/行业数量统计


class SupplyChainCoverageItem(BaseModel):
    """自选股供应链覆盖情况。"""

    code: str
    name: str
    upstream_count: int = 0
    downstream_count: int = 0
    competitor_count: int = 0
    total_count: int = 0
    status: str


class SupplyChainCoverageResponse(BaseModel):
    """供应链模块覆盖率审计。"""

    items: list[SupplyChainCoverageItem]
    stats: dict[str, int]

File: frontend/lib/api/supply_chain.ts
import { api } from "./client";

export interface SupplyChainNode {
  id: number;
  relation_type: "upstream" | "downstream" | "competitor";
  company_name: string;
  company_code: string | null;
  product_desc: string | null;
  percentage: number | null;
  importance: "high" | "medium" | "low";
  is_listed: boolean;
  recent_event_count: number;
  urgent_event_count: number;
  latest_event_title: string | null;
  latest_news_id: number | null;
}

export interface SupplyChainEventSummary {
  news_id: number;
  stock_code: string;
  stock_name: string;
  company_name: string;
  company_code: string | null;
  title: string;
  published_at: string | null;
  urgency: "urgent" | "important" | "info" | null;
  importance_score: number | null;
  relevance: number;
  impact_direction: "bullish" | "bearish" | "neutral" | null;
  impact_summary: string | null;
}

export interface SupplyChain {
  code: string;
  name: string;
  upstream: SupplyChainNode[];
  downstream: SupplyChainNode[];
  competitors: SupplyChainNode[];
  updated_at: string | null;
  recent_events: SupplyChainEventSummary[];
}

// 全局供应链图(聚簇分组)
export interface SupplyChainStockMeta {
  code: string;
  name: string;
  industry: string | null;
  market: string;  // A / H / EXT
  recent_event_count: number;
  urgent_event_count: number;
  latest_event_title: string | null;
  latest_news_id: number | null;
}

export interface SupplyChainEdge {
  from_code: string;
  to_code: string;
  from_name: string;
  to_name: string;
  product_desc: string | null;
  importance: "high" | "medium" | "low";
  relation_type: "upstream" | "downstream";
  both_listed: boolean;
  recent_event_count: number;
  urgent_event_count: number;
  latest_event_title: string | null;
  latest_news_id: number | null;
}

export interface GlobalSupplyChain {
  watchlist_stocks: SupplyChainStockMeta[];
  external_companies: SupplyChainStockMeta[];
  edges: SupplyChainEdge[];
  recent_events: SupplyChainEventSummary[];
  industry_groups: Record<string, string[]>;
  stats: {
    watchlist_count: number;
    external_count: number;
    edge_count: number;
    industry_count: number;
    cross_watchlist_edges: number;
    highlighted_edges: number;
    recent_event_count: number;
  };
}

export const supplyChainApi = {
  get: (code: string) => api.get<SupplyChain>(`/supply-chain/${code}`),
  refresh: (code: string) => api.post(`/supply-chain/${code}/refresh`, {}),
  global: () => api.get<GlobalSupplyChain>("/supply-chain/global"),
  coverage: () => api.get<{
    items: Array<{
      code: string;
      name: string;
      upstream_count: number;
      downstream_count: number;
      competitor_count: number;
      total_count: number;
      status: "complete" | "partial" | "missing";
    }>;
    stats: Record<string, number>;
  }>("/supply-chain/global/coverage"),
  fillGaps: (limit = 10, force = false) =>
    api.post(`/supply-chain/global/fill-gaps?limit=${limit}&force=${force}`, {}),
  refreshIntelligence: (days = 14) =>
    api.post(`/supply-chain/global/refresh-intelligence?days=${days}`, {}),
};
